ICE: `unhandled node Crate(Mod`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_hir_analysis/src/collect/generics_of.rs:226:14: unhandled node Crate(Mod { spans: ModSpans { inner_span: /tmp/im/a.rs:1:1: 16:17 (#0), inject_use_span: /tmp/im/a.rs:2:1: 2:1 (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ a[1fdb]::{use#0}) }, ItemId { owner_id: DefId(0:2 ~ a[1fdb]::std) }, ItemId { owner_id: DefId(0:3 ~ a[1fdb]::Role) }, ItemId { owner_id: DefId(0:5 ~ a[1fdb]::HandshakeCallback) }, ItemId { owner_id: DefId(0:8 ~ a[1fdb]::{impl#0}) }, ItemId { owner_id: DefId(0:11 ~ a[1fdb]::Handshake) }, ItemId { owner_id: DefId(0:14 ~ a[1fdb]::accept) }, ItemId { owner_id: DefId(0:16 ~ a[1fdb]::main) }] })', 'error: internal compiler error: compiler/rustc_hir_analysis/src/collect/generics_of.rs:226:14: unhandled node Crate(Mod { spans: ModSpans { inner_span: /tmp/im/a.rs:1:1: 16:17 (#0), inject_use_span: /tmp/im/a.rs:2:1: 2:1 (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ a[1fdb]::{use#0}) }, ItemId { owner_id: DefId(0:2 ~ a[1fdb]::std) }, ItemId { owner_id: DefId(0:3 ~ a[1fdb]::Role) }, ItemId { owner_id: DefId(0:5 ~ a[1fdb]::HandshakeCallback) }, ItemId { owner_id: DefId(0:8 ~ a[1fdb]::{impl#0}) }, ItemId { owner_id: DefId(0:11 ~ a[1fdb]::Handshake) }, ItemId { owner_id: DefId(0:14 ~ a[1fdb]::accept) }, ItemId { owner_id: DefId(0:16 ~ a[1fdb]::main) }] })'
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
trait Role {
    type Inner = usize;
}
struct HandshakeCallback<C>;
impl<C: FnOnce()> Role for HandshakeCallback {
    type Inner = usize;
}
struct Handshake<R: Role> {
    _inner: Option<R::Inner>,
}
fn accept<C>() -> Handshake<HandshakeCallback<C>> {
    todo!()
}

````

original:
````rust
#![feature(associated_type_defaults)]
trait Role {
    type Inner = usize;
}
struct HandshakeCallback<C>;
impl<C: FnOnce()> Role for HandshakeCallback {
    type Inner = usize;
}
struct Handshake<R: Role> {
    _inner: Option<R::Inner>,
}
fn accept<C>() -> Handshake<HandshakeCallback<C>> {
    todo!()
}

pub fn main() {}

````

Version information
````
rustc 1.90.0-nightly (733b47ea4 2025-07-05)
binary: rustc
commit-hash: 733b47ea4b1b86216f14ef56e49440c33933f230
commit-date: 2025-07-05
host: x86_64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.7

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/733b47ea4b1b86216f14ef56e49440c33933f230/compiler/rustc_hir_analysis/src/collect/generics_of.rs#L220-L232

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: associated type defaults are unstable
 --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:2:5
  |
2 |     type Inner = usize;
  |     ^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #29661 <https://github.com/rust-lang/rust/issues/29661> for more information
  = help: add `#![feature(associated_type_defaults)]` to the crate attributes to enable
  = note: this compiler was built on 2025-07-05; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:13:2
   |
13 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs`

error[E0107]: missing generics for struct `HandshakeCallback`
 --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:5:28
  |
5 | impl<C: FnOnce()> Role for HandshakeCallback {
  |                            ^^^^^^^^^^^^^^^^^ expected 1 generic argument
  |
note: struct defined here, with 1 generic parameter: `C`
 --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:4:8
  |
4 | struct HandshakeCallback<C>;
  |        ^^^^^^^^^^^^^^^^^ -
help: add missing generic argument
  |
5 | impl<C: FnOnce()> Role for HandshakeCallback<C> {
  |                                             +++

error[E0392]: type parameter `C` is never used
 --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:4:26
  |
4 | struct HandshakeCallback<C>;
  |                          ^ unused type parameter
  |
  = help: consider removing `C`, referring to it in a field, or using a marker such as `PhantomData`
  = help: if you intended `C` to be a const parameter, use `const C: /* Type */` instead

error[E0283]: type annotations needed
  --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:11:19
   |
11 | fn accept<C>() -> Handshake<HandshakeCallback<C>> {
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `C`
   |
   = note: multiple `impl`s satisfying `_: FnOnce()` found in the following crates: `alloc`, `core`:
           - impl<A, F> FnOnce<A> for &F
             where A: Tuple, F: Fn<A>, F: ?Sized;
           - impl<A, F> FnOnce<A> for &mut F
             where A: Tuple, F: FnMut<A>, F: ?Sized;
           - impl<Args, F, A> FnOnce<Args> for Box<F, A>
             where Args: Tuple, F: FnOnce<Args>, A: Allocator, F: ?Sized;
           - impl<F, Args> FnOnce<Args> for Exclusive<F>
             where F: FnOnce<Args>, Args: Tuple;
           - impl<R, F> FnOnce<()> for AssertUnwindSafe<F>
             where <F as FnOnce<()>>::Output == R, F: FnOnce();
note: required for `HandshakeCallback<C>` to implement `Role`
  --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:5:19
   |
5  | impl<C: FnOnce()> Role for HandshakeCallback {
   |         --------  ^^^^     ^^^^^^^^^^^^^^^^^
   |         |
   |         unsatisfied trait bound introduced here
note: required by a bound in `Handshake`
  --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:8:21
   |
8  | struct Handshake<R: Role> {
   |                     ^^^^ required by this bound in `Handshake`

error: internal compiler error: compiler/rustc_hir_analysis/src/collect/generics_of.rs:226:14: unhandled node Crate(Mod { spans: ModSpans { inner_span: /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:1:1: 13:2 (#0), inject_use_span: no-location (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ mvce[42ad]::{use#0}) }, ItemId { owner_id: DefId(0:2 ~ mvce[42ad]::std) }, ItemId { owner_id: DefId(0:3 ~ mvce[42ad]::Role) }, ItemId { owner_id: DefId(0:5 ~ mvce[42ad]::HandshakeCallback) }, ItemId { owner_id: DefId(0:8 ~ mvce[42ad]::{impl#0}) }, ItemId { owner_id: DefId(0:11 ~ mvce[42ad]::Handshake) }, ItemId { owner_id: DefId(0:14 ~ mvce[42ad]::accept) }] })
  --> /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:1:1
   |
1  | / trait Role {
2  | |     type Inner = usize;
3  | | }
4  | | struct HandshakeCallback<C>;
...  |
12 | |     todo!()
13 | | }
   | |_^


thread 'rustc' panicked at compiler/rustc_hir_analysis/src/collect/generics_of.rs:226:14:
Box<dyn Any>
stack backtrace:
   0:     0x7c1807536883 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::ha2e7e3a01df69042
   1:     0x7c1807c0273f - core::fmt::write::h7cb8f63788cd01d2
   2:     0x7c180752c4c3 - std::io::Write::write_fmt::h5ebaf381f8338068
   3:     0x7c18075366e2 - std::sys::backtrace::BacktraceLock::print::h3ec4d7883eb25e61
   4:     0x7c180753a17a - std::panicking::default_hook::{{closure}}::h29548987efd832cb
   5:     0x7c1807539cff - std::panicking::default_hook::ha25170a15c643514
   6:     0x7c18066056a3 - std[d534d36ffd607ff3]::panicking::update_hook::<alloc[756c202a2770bd7e]::boxed::Box<rustc_driver_impl[4ab48184b7847074]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7c180753a9cb - std::panicking::rust_panic_with_hook::h0d81afcd829aa24b
   8:     0x7c180663f3b1 - std[d534d36ffd607ff3]::panicking::begin_panic::<rustc_errors[395af0118e3b23d6]::ExplicitBug>::{closure#0}
   9:     0x7c18066338c6 - std[d534d36ffd607ff3]::sys::backtrace::__rust_end_short_backtrace::<std[d534d36ffd607ff3]::panicking::begin_panic<rustc_errors[395af0118e3b23d6]::ExplicitBug>::{closure#0}, !>
  10:     0x7c1806633757 - std[d534d36ffd607ff3]::panicking::begin_panic::<rustc_errors[395af0118e3b23d6]::ExplicitBug>
  11:     0x7c1806648eb1 - <rustc_errors[395af0118e3b23d6]::diagnostic::BugAbort as rustc_errors[395af0118e3b23d6]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7c1806b89f9c - <rustc_errors[395af0118e3b23d6]::DiagCtxtHandle>::span_bug::<rustc_span[439cc0b179e5ffc0]::span_encoding::Span, alloc[756c202a2770bd7e]::string::String>
  13:     0x7c1806c1eb87 - rustc_middle[1c3bc8d0cdd03ead]::util::bug::opt_span_bug_fmt::<rustc_span[439cc0b179e5ffc0]::span_encoding::Span>::{closure#0}
  14:     0x7c1806bfec4a - rustc_middle[1c3bc8d0cdd03ead]::ty::context::tls::with_opt::<rustc_middle[1c3bc8d0cdd03ead]::util::bug::opt_span_bug_fmt<rustc_span[439cc0b179e5ffc0]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x7c1806bfeabb - rustc_middle[1c3bc8d0cdd03ead]::ty::context::tls::with_context_opt::<rustc_middle[1c3bc8d0cdd03ead]::ty::context::tls::with_opt<rustc_middle[1c3bc8d0cdd03ead]::util::bug::opt_span_bug_fmt<rustc_span[439cc0b179e5ffc0]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x7c18047927f7 - rustc_middle[1c3bc8d0cdd03ead]::util::bug::span_bug_fmt::<rustc_span[439cc0b179e5ffc0]::span_encoding::Span>
  17:     0x7c18050dbbc0 - rustc_hir_analysis[a54d78a35e7f2910]::collect::generics_of::generics_of
  18:     0x7c1807c193a9 - rustc_query_impl[65f50324650e1fa3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[65f50324650e1fa3]::query_impl::generics_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 8usize]>>
  19:     0x7c1807c19aa5 - rustc_query_system[2bb68dc6eee57125]::query::plumbing::try_execute_query::<rustc_query_impl[65f50324650e1fa3]::DynamicConfig<rustc_query_system[2bb68dc6eee57125]::query::caches::DefIdCache<rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[65f50324650e1fa3]::plumbing::QueryCtxt, false>
  20:     0x7c1807c197dd - rustc_query_impl[65f50324650e1fa3]::query_impl::generics_of::get_query_non_incr::__rust_end_short_backtrace
  21:     0x7c1808a2c65a - rustc_hir_analysis[a54d78a35e7f2910]::collect::predicates_of::gather_explicit_predicates_of::{closure#0}
  22:     0x7c1808a2a22a - rustc_query_impl[65f50324650e1fa3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[65f50324650e1fa3]::query_impl::explicit_predicates_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 24usize]>>
  23:     0x7c1807d956a6 - rustc_query_system[2bb68dc6eee57125]::query::plumbing::try_execute_query::<rustc_query_impl[65f50324650e1fa3]::DynamicConfig<rustc_query_system[2bb68dc6eee57125]::query::caches::DefIdCache<rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[65f50324650e1fa3]::plumbing::QueryCtxt, false>
  24:     0x7c1807d94fe2 - rustc_query_impl[65f50324650e1fa3]::query_impl::explicit_predicates_of::get_query_non_incr::__rust_end_short_backtrace
  25:     0x7c1807d943be - rustc_query_impl[65f50324650e1fa3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[65f50324650e1fa3]::query_impl::predicates_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 24usize]>>
  26:     0x7c1807d956c5 - rustc_query_system[2bb68dc6eee57125]::query::plumbing::try_execute_query::<rustc_query_impl[65f50324650e1fa3]::DynamicConfig<rustc_query_system[2bb68dc6eee57125]::query::caches::DefIdCache<rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[65f50324650e1fa3]::plumbing::QueryCtxt, false>
  27:     0x7c1807d94ee2 - rustc_query_impl[65f50324650e1fa3]::query_impl::predicates_of::get_query_non_incr::__rust_end_short_backtrace
  28:     0x7c18073522f3 - rustc_trait_selection[4fadcde6e70bd94c]::error_reporting::traits::ambiguity::compute_applicable_impls_for_diagnostics
  29:     0x7c18073e6cde - <rustc_trait_selection[4fadcde6e70bd94c]::error_reporting::TypeErrCtxt>::maybe_report_ambiguity
  30:     0x7c18073b8fad - <rustc_trait_selection[4fadcde6e70bd94c]::error_reporting::TypeErrCtxt>::report_fulfillment_errors
  31:     0x7c18084a9a8a - <rustc_borrowck[6828e82344ae7e12]::type_check::liveness::trace::LivenessContext>::compute_drop_data
  32:     0x7c1804ba60f3 - rustc_borrowck[6828e82344ae7e12]::type_check::liveness::trace::trace
  33:     0x7c180502fd0d - rustc_borrowck[6828e82344ae7e12]::type_check::type_check
  34:     0x7c180909cf49 - rustc_borrowck[6828e82344ae7e12]::do_mir_borrowck
  35:     0x7c1807dce6ad - rustc_borrowck[6828e82344ae7e12]::mir_borrowck
  36:     0x7c1807dce3d9 - rustc_query_impl[65f50324650e1fa3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[65f50324650e1fa3]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 8usize]>>
  37:     0x7c1807dcd4ae - rustc_query_system[2bb68dc6eee57125]::query::plumbing::try_execute_query::<rustc_query_impl[65f50324650e1fa3]::DynamicConfig<rustc_data_structures[64cba4f9a751aff]::vec_cache::VecCache<rustc_span[439cc0b179e5ffc0]::def_id::LocalDefId, rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[2bb68dc6eee57125]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[65f50324650e1fa3]::plumbing::QueryCtxt, false>
  38:     0x7c1807dcccc9 - rustc_query_impl[65f50324650e1fa3]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  39:     0x7c180860393c - <rustc_middle[1c3bc8d0cdd03ead]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface[331c37c46c9c8d25]::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
  40:     0x7c1808601382 - rustc_interface[331c37c46c9c8d25]::passes::analysis
  41:     0x7c180860082b - rustc_query_impl[65f50324650e1fa3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[65f50324650e1fa3]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 0usize]>>
  42:     0x7c1808c88c7c - rustc_query_system[2bb68dc6eee57125]::query::plumbing::try_execute_query::<rustc_query_impl[65f50324650e1fa3]::DynamicConfig<rustc_query_system[2bb68dc6eee57125]::query::caches::SingleCache<rustc_middle[1c3bc8d0cdd03ead]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[65f50324650e1fa3]::plumbing::QueryCtxt, false>
  43:     0x7c1808c88846 - rustc_query_impl[65f50324650e1fa3]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  44:     0x7c1808e8a9d6 - rustc_interface[331c37c46c9c8d25]::passes::create_and_enter_global_ctxt::<core[a1c5ac6c25c163e6]::option::Option<rustc_interface[331c37c46c9c8d25]::queries::Linker>, rustc_driver_impl[4ab48184b7847074]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  45:     0x7c1808e58c03 - rustc_interface[331c37c46c9c8d25]::interface::run_compiler::<(), rustc_driver_impl[4ab48184b7847074]::run_compiler::{closure#0}>::{closure#1}
  46:     0x7c1808df19f8 - std[d534d36ffd607ff3]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[331c37c46c9c8d25]::util::run_in_thread_with_globals<rustc_interface[331c37c46c9c8d25]::util::run_in_thread_pool_with_globals<rustc_interface[331c37c46c9c8d25]::interface::run_compiler<(), rustc_driver_impl[4ab48184b7847074]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  47:     0x7c1808df16d6 - <<std[d534d36ffd607ff3]::thread::Builder>::spawn_unchecked_<rustc_interface[331c37c46c9c8d25]::util::run_in_thread_with_globals<rustc_interface[331c37c46c9c8d25]::util::run_in_thread_pool_with_globals<rustc_interface[331c37c46c9c8d25]::interface::run_compiler<(), rustc_driver_impl[4ab48184b7847074]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[a1c5ac6c25c163e6]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  48:     0x7c1808defd3d - std::sys::pal::unix::thread::Thread::new::thread_start::h1a22ded422ce395b
  49:     0x7c180285f7eb - <unknown>
  50:     0x7c18028e318c - <unknown>
  51:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.90.0-nightly (733b47ea4 2025-07-05) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [generics_of] computing generics of ``
#1 [explicit_predicates_of] computing explicit predicates of ``
#2 [predicates_of] computing predicates of ``
#3 [mir_borrowck] borrow-checking `accept`
#4 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 6 previous errors

Some errors have detailed explanations: E0107, E0283, E0392, E0601, E0658.
For more information about an error, try `rustc --explain E0107`.

```

</p>
</details>

<!--
query stack:
error: internal compiler error: compiler/rustc_hir_analysis/src/collect/generics_of.rs:226:14: unhandled node Crate(Mod { spans: ModSpans { inner_span: /tmp/icemaker_global_tempdir.C27Sk6SxxVwd/rustc_testrunner_tmpdir_reporting.fP7xhS12mjRQ/mvce.rs:1:1: 13:2 (#0), inject_use_span: no-location (#0) }, item_ids: [ItemId { owner_id: DefId(0:1 ~ mvce[42ad]::{use#0}) }, ItemId { owner_id: DefId(0:2 ~ mvce[42ad]::std) }, ItemId { owner_id: DefId(0:3 ~ mvce[42ad]::Role) }, ItemId { owner_id: DefId(0:5 ~ mvce[42ad]::HandshakeCallback) }, ItemId { owner_id: DefId(0:8 ~ mvce[42ad]::{impl#0}) }, ItemId { owner_id: DefId(0:11 ~ mvce[42ad]::Handshake) }, ItemId { owner_id: DefId(0:14 ~ mvce[42ad]::accept) }] })
#0 [generics_of] computing generics of ``
#1 [explicit_predicates_of] computing explicit predicates of ``
#2 [predicates_of] computing predicates of ``
#3 [mir_borrowck] borrow-checking `accept`
#4 [analysis] running analysis passes on this crate

-->
@rustbot label +F-associated_type_defaults

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/ambiguity.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/ambiguity.rs
@@ -4,7 +4,7 @@ use rustc_errors::{Applicability, Diag, E0283, E0284, E0790, MultiSpan, struct_s
 use rustc_hir as hir;
 use rustc_hir::LangItem;
 use rustc_hir::def::{DefKind, Res};
-use rustc_hir::def_id::DefId;
+use rustc_hir::def_id::{CRATE_DEF_ID, DefId};
 use rustc_hir::intravisit::Visitor as _;
 use rustc_infer::infer::{BoundRegionConversionTime, InferCtxt};
 use rustc_infer::traits::util::elaborate;
@@ -128,19 +128,26 @@ pub fn compute_applicable_impls_for_diagnostics<'tcx>(
         },
     );
 
-    let predicates =
-        tcx.predicates_of(obligation.cause.body_id.to_def_id()).instantiate_identity(tcx);
-    for (pred, span) in elaborate(tcx, predicates.into_iter()) {
-        let kind = pred.kind();
-        if let ty::ClauseKind::Trait(trait_pred) = kind.skip_binder()
-            && param_env_candidate_may_apply(kind.rebind(trait_pred))
-        {
-            if kind.rebind(trait_pred.trait_ref)
-                == ty::Binder::dummy(ty::TraitRef::identity(tcx, trait_pred.def_id()))
+    // If our `body_id` has been set (and isn't just from a dummy obligation cause),
+    // then try to look for a param-env clause that would apply. The way we compute
+    // this is somewhat manual, since we need the spans, so we elaborate this directly
+    // from `predicates_of` rather than actually looking at the param-env which
+    // otherwise would be more appropriate.
+    let body_id = obligation.cause.body_id;
+    if body_id != CRATE_DEF_ID {
+        let predicates = tcx.predicates_of(body_id.to_def_id()).instantiate_identity(tcx);
+        for (pred, span) in elaborate(tcx, predicates.into_iter()) {
+            let kind = pred.kind();
+            if let ty::ClauseKind::Trait(trait_pred) = kind.skip_binder()
+                && param_env_candidate_may_apply(kind.rebind(trait_pred))
             {
-                ambiguities.push(CandidateSource::ParamEnv(tcx.def_span(trait_pred.def_id())))
-            } else {
-                ambiguities.push(CandidateSource::ParamEnv(span))
+                if kind.rebind(trait_pred.trait_ref)
+                    == ty::Binder::dummy(ty::TraitRef::identity(tcx, trait_pred.def_id()))
+                {
+                    ambiguities.push(CandidateSource::ParamEnv(tcx.def_span(trait_pred.def_id())))
+                } else {
+                    ambiguities.push(CandidateSource::ParamEnv(span))
+                }
             }
         }
     }

--- a/tests/ui/traits/error-reporting/ambiguity-in-dropck-err-reporting.rs
+++ b/tests/ui/traits/error-reporting/ambiguity-in-dropck-err-reporting.rs
@@ -1,0 +1,18 @@
+// Regression test for #143481, where we were calling `predicates_of` on
+// a Crate HIR node because we were using a dummy obligation cause's body id
+// without checking that it was meaningful first.
+
+trait Role {
+    type Inner;
+}
+struct HandshakeCallback<C>(C);
+impl<C: Clone> Role for HandshakeCallback {
+    //~^ ERROR missing generics
+    type Inner = usize;
+}
+struct Handshake<R: Role>(R::Inner);
+fn accept() -> Handshake<HandshakeCallback<()>> {
+    todo!()
+}
+
+fn main() {}

--- a/tests/ui/traits/error-reporting/ambiguity-in-dropck-err-reporting.stderr
+++ b/tests/ui/traits/error-reporting/ambiguity-in-dropck-err-reporting.stderr
@@ -1,0 +1,19 @@
+error[E0107]: missing generics for struct `HandshakeCallback`
+  --> $DIR/ambiguity-in-dropck-err-reporting.rs:9:25
+   |
+LL | impl<C: Clone> Role for HandshakeCallback {
+   |                         ^^^^^^^^^^^^^^^^^ expected 1 generic argument
+   |
+note: struct defined here, with 1 generic parameter: `C`
+  --> $DIR/ambiguity-in-dropck-err-reporting.rs:8:8
+   |
+LL | struct HandshakeCallback<C>(C);
+   |        ^^^^^^^^^^^^^^^^^ -
+help: add missing generic argument
+   |
+LL | impl<C: Clone> Role for HandshakeCallback<C> {
+   |                                          +++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0107`.

